actionpack/lib/action_controller/metal/rendering.rb

```diff
@@ -4,6 +4,17 @@ module Rendering
 
     RENDER_FORMATS_IN_PRIORITY = [:body, :text, :plain, :html]
 
+    module ClassMethods
+      # Documentation at ActionController::Renderer#render
+      delegate :render, to: :renderer
+
+      # Returns a renderer class (inherited from ActionController::Renderer)
+      # for the controller.
+      def renderer
+        @renderer ||= Renderer.for(self)
```

Comment:
Right, which is why there's a thread safety issue. `Renderer.for(ApplicationController)` should always return the same instance, which it doesn't in this case. If we have any mutable state in the renderer which we'd like to rely on, this will break.